README.md
# php-binarytranscoder

[![GitHub release](https://img.shields.io/github/release/ryanwhowe/php-binarytranscoder.svg)](https://github.com/ryanwhowe/php-binarytranscoder)

### WARNING!!!

#### This Is a Breaking Change from the v1 branch

The encoder and decoder have been changed to reverse the binary string that transcoded from the input array.  This will 
allow for bitwise searches of the encoded integer values in a backwards compatible way.  This will cause a breaking 
change from the v1 branch if this is used to decode any integers encoded by the v1 package.  Any information encoded by
the v1 package will need to be recoded.

Convert integer values to binary arrays in a backwards compatible safe method

While that might sound like a lot it is simply storing a php array of boolean values in an integer.  This integer can be
stored in a database or by some other means.  The array of booleans can be added to and it will be backwards compatible 
with the stored integer representation of the boolean array.

### Warning:
If you use this to store information into a database, this is stored in a nice compact form, however you need to keep in 
mind that it is stored in a form that make direct querying very difficult.  For example, if you stored a set of user 
permissions as an integer, it would be difficult to directly query the database for a list of users who have a given 
permission 'X' .

## Basic Usage
### Encoding
To encoding an array is to transform the representation of the array of boolean values to an integer representation for storage.
```php
$raw_array = array(
    'value 1' => true,
    'value 2' => false,
    'value 3' => true
);

$transcoder = new \ryanwhowe\BinaryTranscoder($raw_array);
$encoded_value = $transcoder->encodeArray($raw_array);
var_dump($encoded_value);
```
output
```text
int(13)
```
### Decoding
decoding a value back to an array
```php
use RyanWHowe\BinaryTranscoder\BinaryTranscoder;
$output_array_key_values = array(
    'value 1',
    'value 2',
    'value 3'
);
$encoded_stored_value = 13;

$transcoder = new BinaryTranscoder($output_array_key_values);
$result = $transcoder->decodeInteger($encoded_stored_value);
var_dump($result);

```
output
```text
array(3) {
  'value 1' => bool(true)
  'value 2' => bool(false)
  'value 3' => bool(true)
}
```

### Max Integer Value (new v2.0)
If storing the trancoded integer value an implementation may need to limit the maximum allowable integer value to 
something other than the default PHP_INT_MAX value.  **Important:** you can **NOT** exceed the PHP_INT_MAX value as this 
is a limitation of PHP.  The example below limits to an unsigned 32 bit integer value (the max for a MySQL UNSIGNED INT column)
```php
use RyanWHowe\BinaryTranscoder\BinaryTranscoder;
$transcoder = new BinaryTranscoder($array_keys, 4294967295);
```
### Padding
The default behavior of the transcoder is to have any newly added array key default to false if there was no encoded 
value for that key.  This behavior can be altered to default to true when instantiating the object, or to null.
```php
use RyanWHowe\BinaryTranscoder\BinaryTranscoder;
$transcoder = new BinaryTranscoder($array_keys, PHP_INT_MAX, \ryanwhowe\BinaryTranscoder::BOOLEAN_PAD_TRUE);
```

## Advanced Usage
### Max Int
The class will check the system **PHP_INT_MAX** size when instantiated.  This is through the static method 
_determineMaxArrayLength()_ which accepts an optional integer parameter to allow you to determine the columns that can be
encoded from a source array.  This utility function can be used with the documented limits of a particular database 
storage type to ensure you are not exceeding you encoding abilities.  In general you will always have 1 fewer storage 
positions than you have bits in the integer that is being encoded (see [methodology](#methodology) for more detail).

As of version 2.0 the max integer value can be set to something lower than **PHP_INT_MAX**.  This can be useful if 
storing the value in a 32bit integer, i.e. MySQL INT

#### Warning
Regardless of the storage size limitations you can not exceed your systems **PHP_INT_MAX** size!

#### example
```php
use RyanWHowe\BinaryTranscoder\BinaryTranscoder;
$mysql_small_unsigned_int_max = 65535;
$max_columns = BinaryTranscoder::determineMaxArrayLength($mysql_small_unsigned_int_max);
var_dump($max_columns);
```
output
```text
int(15)
```

## Methodology
The information is translated from a boolean array to a binary string converting true values to 1 and false values to 0.
In order to ensure the decoding process there is an additional most significant bit added to the beginning of the binary 
string.  This string is also reversed so that the oldest values are at the end of the string.  This is for bitwise 
searching of the data.
````text
[false, false, true, false] would become 10100 NOT 0100 
````
This creates a minimum value that will be stored, even for a completely false set of array values, which will be a 
constant for the count of the array being used.  This also provides a "versioning" of the stored integer value.

The best use case for the class is to have it be a resource of another class that has control over the array that will 
be encoded and decoded into.


CHANGELOG.md
V2.0.0 ChangeLog
* Updated minimal requirements to php 7.0.0 for strict types
* Added CHANGELOG.md
* Added UPDATE_2.0.md
* Addded Custom Exception \ryanwhowe\BinaryTranscoderException
* Addded Custom Exception \ryanwhowe\BinaryTranscoderMaxLengthException
* Addded Custom Exception \ryanwhowe\BinaryTranscoderIntegerOverflowException
* BinaryTranscoder::__construct now throws a \ryanwhowe\BinaryTranscoderMaxLengthException when the number of elements that can be trasncoded exceeds the system amount available in an integer.
* BinaryTranscoder::__construct now throws a \ryanwhowe\BinaryTranscoderIntegerOverflowException when the maximum integer value parameter is set higher than PHP_INT_MAX.
* BinaryTranscoder::__construct now takes an optional maximum integer value parameter.
* BinaryTranscoder::__construct padding paramater has been moved to the third position after the maximum integer value parameter.
* BinaryTranscoder.decodeInteger() will now throw \ryanwhowe\BinaryTranscoderException when the passed integer is to small or two large for the key array used on instantiation.
* BinaryTranscoder.encodeArray() will now throw \ryanwhowe\BinaryTranscoderException when the passed array is of different length than the key array used on instantiation.
* Implemented strict types for all methods
* Updated tests for reverse encoding method
* Namespace changed from ryanwhowe to RyanWHowe\BinaryTranscoder
* Changed the package name from ryanwhowe/php-binarytranscoder to ryanwhowe/binarytranscoder


test/BinaryTranscoderTest.php
<?php
/**
 * This file contains the definition for the BinaryTranscoderTest class
 *
 * @author Ryan Howe
 * @since  2018-09-29
 */
namespace Test;

use RyanWHowe\BinaryTranscoder\BinaryTranscoder;

class BinaryTranscoderTest extends \PHPUnit\Framework\TestCase
{

    /**
     * A dataProvider that has data in the form of $encodedValue, $expectedDecodedValue
     *
     * @return array
     */
    public function transcodingProvider()
    {
        return [
            [ bindec('1 000'), ['key1' => false, 'key2' => false, 'key3' => false] ],
            [ bindec('1 001'), ['key1' => true,  'key2' => false, 'key3' => false] ],
            [ bindec('1 010'), ['key1' => false, 'key2' => true,  'key3' => false] ],
            [ bindec('1 100'), ['key1' => false, 'key2' => false, 'key3' => true] ],
            [ bindec('1 011'), ['key1' => true,  'key2' => true,  'key3' => false] ],
            [ bindec('1 110'), ['key1' => false, 'key2' => true,  'key3' => true] ],
            [ bindec('1 111'), ['key1' => true,  'key2' => true,  'key3' => true] ],
        ];
    }

    /**
     * A dataProvider that has data in the form of $encodedStoredValue, $expectedDecodedValue .. These
     * expectedDecodedValues and what are expected when the Padding_Boolean is set to false
     *
     * @return array
     */
    public function forwardTranscodingFalsePadProvider()
    {
        return [
            [ bindec('1 000'), ['key1' => false, 'key2' => false, 'key3' => false, 'key4' => false] ],
            [ bindec('1 001'), ['key1' => true,  'key2' => false, 'key3' => false, 'key4' => false] ],
            [ bindec('1 010'), ['key1' => false, 'key2' => true,  'key3' => false, 'key4' => false] ],
            [ bindec('1 100'), ['key1' => false, 'key2' => false, 'key3' => true,  'key4' => false] ],
            [ bindec('1 011'), ['key1' => true,  'key2' => true,  'key3' => false, 'key4' => false] ],
            [ bindec('1 110'), ['key1' => false, 'key2' => true,  'key3' => true,  'key4' => false] ],
            [ bindec('1 111'), ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => false] ],
            [
                bindec('1111'),
                ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => false, 'key5' => false, 'key6' => false]
            ],
        ];
    }

    /**
     * A dataProvider that has data in the form of $encodedStoredValue, $expectedDecodedValue .. These
     * expectedDecodedValues and what are expected when the Padding_Boolean is set to true
     *
     * @return array
     */
    public function forwardTranscodingTruePadProvider()
    {
        return [
            [ bindec('1 000'), ['key1' => false, 'key2' => false, 'key3' => false, 'key4' => true] ],
            [ bindec('1 001'), ['key1' => true,  'key2' => false, 'key3' => false, 'key4' => true] ],
            [ bindec('1 010'), ['key1' => false, 'key2' => true,  'key3' => false, 'key4' => true] ],
            [ bindec('1 100'), ['key1' => false, 'key2' => false, 'key3' => true,  'key4' => true] ],
            [ bindec('1 011'), ['key1' => true,  'key2' => true,  'key3' => false, 'key4' => true] ],
            [ bindec('1 110'), ['key1' => false, 'key2' => true,  'key3' => true,  'key4' => true] ],
            [ bindec('1 111'), ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => true] ],
            [
                bindec('1111'),
                ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => true, 'key5' => true, 'key6' => true]
            ],
        ];
    }

    /**
     * A dataProvider that has data in the form of $encodedStoredValue, $expectedDecodedValue .. These
     * expectedDecodedValues and what are expected when the Padding_Boolean is set to null
     *
     * @return array
     */
    public function forwardTranscodingNullPadProvider()
    {
        return [
            [ bindec('1 000'), ['key1' => false, 'key2' => false, 'key3' => false, 'key4' => null] ],
            [ bindec('1 001'), ['key1' => true,  'key2' => false, 'key3' => false, 'key4' => null] ],
            [ bindec('1 010'), ['key1' => false, 'key2' => true,  'key3' => false, 'key4' => null] ],
            [ bindec('1 100'), ['key1' => false, 'key2' => false, 'key3' => true,  'key4' => null] ],
            [ bindec('1 011'), ['key1' => true,  'key2' => true,  'key3' => false, 'key4' => null] ],
            [ bindec('1 110'), ['key1' => false, 'key2' => true,  'key3' => true,  'key4' => null] ],
            [ bindec('1 111'), ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => null] ],
            [
                bindec('1111'),
                ['key1' => true,  'key2' => true,  'key3' => true,  'key4' => null, 'key5' => null, 'key6' => null]
            ],
        ];
    }

    /**
     * @test
     * @dataProvider transcodingProvider
     * @dataProvider forwardTranscodingFalsePadProvider
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::decodeInteger()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertStringToArray()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertProtectedIntToBin()
     * @param $encoded_int
     * @param $expected_decode
     * @throws \Exception
     */
    public function decodingFalsePad($encoded_int, $expected_decode)
    {
        $key_array = $expected_decode;
        array_walk(
            $key_array,
            function (&$item) {
                $item = null;
            }
        );
        $tr = new BinaryTranscoder($expected_decode);
        $decode = $tr->decodeInteger($encoded_int);

        $this->assertEquals($expected_decode, $decode);
    }

    /**
     * @test
     * @dataProvider transcodingProvider
     * @dataProvider forwardTranscodingTruePadProvider
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::decodeInteger()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertProtectedIntToBin()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertStringToArray()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength()
     * @param $encoded_int
     * @param $expected_decode
     * @throws \Exception
     */
    public function decodingTruePad($encoded_int, $expected_decode)
    {

        $key_array = $expected_decode;
        array_walk(
            $key_array,
            function (&$item) {
                $item = null;
            }
        );
        $tr = new BinaryTranscoder($key_array, PHP_INT_MAX, BinaryTranscoder::BINARYTRANSCODER_PAD_TRUE);
        $decode = $tr->decodeInteger($encoded_int);

        $this->assertEquals($expected_decode, $decode);
    }

    /**
     * @test
     * @dataProvider transcodingProvider
     * @dataProvider forwardTranscodingNullPadProvider
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::decodeInteger()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertProtectedIntToBin()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertStringToArray()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength()
     * @param $encoded_int
     * @param $expected_decode
     * @throws \Exception
     */
    public function decodingNullPad($encoded_int, $expected_decode)
    {

        $key_array = $expected_decode;
        array_walk(
            $key_array,
            function (&$item) {
                $item = null;
            }
        );
        $tr = new BinaryTranscoder($key_array, PHP_INT_MAX, BinaryTranscoder::BINARYTRANSCODER_PAD_NULL);
        $decode = $tr->decodeInteger($encoded_int);

        $this->assertEquals($expected_decode, $decode);
    }

    /**
     * @test
     * @dataProvider transcodingProvider
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::encodeArray()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertArrayToString()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertBinToProtectedInt()
     * @covers       \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength()
     * @param $expected_encode_int
     * @param $test_array
     * @throws \Exception
     */
    public function encoding($expected_encode_int, $test_array)
    {

        $tr = new BinaryTranscoder($test_array);
        $encoded = $tr->encodeArray($test_array);

        $this->assertEquals($expected_encode_int, $encoded);
    }

    /**
     * @test
     * @expectedException \RyanWHowe\BinaryTranscoder\BinaryTranscoderMaxLengthException
     * @covers \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct()
     * @covers \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength
     * @throws \Exception
     */
    public function maxItemCountOverFlow()
    {
        $array_count = strlen(decbin(PHP_INT_MAX));
        $test_array = array();
        for ($i = 1; $i <= $array_count; $i++) {
            $test_array[$i] = false;
        }
        new BinaryTranscoder($test_array);
    }

    /**
     * @test
     * @dataProvider             transcodingProvider
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::encodeArray()
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertArrayToString
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength
     * @param $expected_encode_int
     * @param $test_array
     * @expectedException \RyanWHowe\BinaryTranscoder\BinaryTranscoderException
     * @expectedExceptionMessage The source array has a different length than the output string!
     * @throws \Exception
     */
    public function itemCountMissMatchEncoder($expected_encode_int, $test_array)
    {
        $tr = new BinaryTranscoder($test_array);
        array_push($test_array, array('key_z' => false));
        $tr->encodeArray($test_array);
    }

    /**
     * @test
     * @dataProvider             transcodingProvider
     * @dataProvider             forwardTranscodingFalsePadProvider
     * @expectedException \RyanWHowe\BinaryTranscoder\BinaryTranscoderException
     * @expectedExceptionMessage The key array does not have enough elements for the decoded integer
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::decodeInteger()
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::__construct
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertProtectedIntToBin
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::convertStringToArray
     * @covers                   \RyanWHowe\BinaryTranscoder\BinaryTranscoder::determineMaxArrayLength
     * @param $encoded_int
     * @param $expected_decode
     * @throws \Exception
     */
    public function itemCountMissMatchDecoder($encoded_int, $expected_decode)
    {
        $key_array = array_keys($expected_decode);
        $tr = new BinaryTranscoder($key_array);
        $failure_decode = bindec(str_pad('1', count($key_array) + 2, '0', STR_PAD_RIGHT));
        $decode = $tr->decodeInteger($failure_decode);
        $this->assertEquals($expected_decode, $decode);
    }
}


src/BinaryTranscoder.php
<?php declare(strict_types=1);
/**
 * This file contains the code for the BinaryTrascoder Class
 *
 * PHP 7.0.0
 * 
 * @since 2018-09-28
 * @author Ryan Howe
 * @version 2.0.0
 */

namespace RyanWHowe\BinaryTranscoder;

/**
 * Class BinaryTranscoder
 *
 * This class is for encoding and decoding integer values into a binary sequence for representing an array
 * of boolean values.  There are certain steps that need to be taken to ensure that backward compatibility
 * will be maintained as new values are added to the array of booleans it will simply default the values to
 * the pad_boolean value (default to false) for anything that is newly added to the bottom of the field array
 *
 * @author Ryan Howe
 * @since 2018-09-29
 * @package ryanwhowe/php-binarytranscoder
 */
class BinaryTranscoder
{

    /**
     * Pad the unmapped key values to false
     */
    const BINARYTRANSCODER_PAD_FALSE = false;

    /**
     * Pad the unmapped key values to true
     */
    const BINARYTRANSCODER_PAD_TRUE = true;

    /**
     * Pad the unmapped key values to null
     */
    const BINARYTRANSCODER_PAD_NULL = null;

    /**
     * @var array The array of ordered keys that will be associated with the binary output
     */
    private $key_array;

    /**
     * @var integer the number of elements in the key array, used to determine the total binary string length used
     */
    private $array_length;

    /**
     * @var boolean Determines if the backwards compatibility is to pad new entries defaulted to true or to false,
     * default to false
     */
    private $pad_boolean;

    /**
     * BinaryTranscoder constructor.
     *
     * @param array $key_array The array of ordered keys that will be associated with the binary output
     * @param int $max_int_value The maximum size that the class should allow
     * @param boolean|null $pad_boolean  The default backwards compatibility value when new array elements are added
     * @throws BinaryTranscoderMaxLengthException if more array elements than what can be transcoded are passes to the constructor
     * @throws BinaryTranscoderIntegerOverflowException if value above PHP_INT_MAP used for $max_int_value
     */
    public function __construct(array $key_array, int $max_int_value = PHP_INT_MAX, $pad_boolean = self::BINARYTRANSCODER_PAD_FALSE)
    {
        $this->key_array = $this->getKeyArray($key_array);

        $count = count($this->key_array);
        if($max_int_value > PHP_INT_MAX){
            throw new BinaryTranscoderIntegerOverflowException("A value larger than PHP_INT_MAX can not be used be the transcoder");
        }
        $max_count = self::determineMaxArrayLength($max_int_value);
        if ($count > $max_count) {
            throw new BinaryTranscoderMaxLengthException("The maximum amount of array field that can be transcoded is 
            {$max_count}");
        }
        $this->array_length = $count;
        if (null === $pad_boolean) {
            $this->pad_boolean = null;
        } else {
            $this->pad_boolean = (boolean)$pad_boolean;
        }
    }

    /**
     * Get the Keys from the passed array unless this is a array of keys itself
     * 
     * @param array $keyArray 
     * @return array
     */
    private function getKeyArray(array $keyArray): array
    {
        if ($this->hasStringKeys($keyArray)) {
            return array_keys($keyArray);
        } else {
            return $keyArray;
        }
    }

    /**
     * Determin if this array has string keys or not, if the keys are integers it is assumed that a array_keys() array was passed
     * 
     * @param array $array The input array to check the keys of to see if they are strings
     * @return bool
     */
    private function hasStringKeys(array $array): bool
    {
        return count(array_filter(array_keys($array), 'is_string')) > 0;
    }

    /**
     * Return the array representation of the decoded protected integer value
     *
     * @param $integer
     * @return array
     * @throws BinaryTranscoderException when the decoded integer does not match the array length generated
     */
    public function decodeInteger(int $integer): array
    {
        $decoded = $this->convertProtectedIntToBin($integer);
        $decoded_array = $this->convertStringToArray($decoded);
        return $decoded_array;
    }

    /**
     * Return the protected integer representation of the passed array
     *
     * @return integer
     * @param array $array_values
     * @throws BinaryTranscoderException when the passed array of values is of different length than the instantiated key array
     */
    public function encodeArray(array $array_values): int
    {
        $result = $this->convertBinToProtectedInt($this->convertArrayToString($array_values));
        return $result;
    }

    /**
     * This function will convert a protected int into a binary representation.  The protected integers
     * are encoded with a leading 1 to preserver the true length of the binary string.  This converter will
     * back fill the binary string with 0 values to the current length of the key array for the class as it was
     * instantiated.
     *
     * @param $source_int integer
     * @return string
     */
    private function convertProtectedIntToBin(int $source_int): string
    {
        $result = decbin($source_int);
        $result = substr($result, 1, strlen($result));
        $result = strrev($result);
        if (null === $this->pad_boolean) {
            return $result;
        } else {
            $pad_string = ($this->pad_boolean) ? '1' : '0';
            return str_pad($result, $this->array_length, $pad_string, STR_PAD_RIGHT);
        }
    }

    /**
     * This function will convert the binary to a protected int.  The protected Int has a leading 1 to
     * preserve the full original length of the array and all the values withing it discretely.  Without
     * the leading 0 the value of 000010 would be stored as 2 and decode to 10.  This protection
     * is necessary for the backwards compatibility of this class.  Without it we would not be able
     * to append new values to the end of the key array and have everything work correctly.  This
     * effectively creates a minimum value that can be stored for a given number of array elements.
     *
     * @param $source_string string the source binary string
     * @return number
     */
    private function convertBinToProtectedInt(string $source_string): int
    {
        $pad_string = ($this->pad_boolean) ? '1' : '0';
        $protected_binary = '1' . str_pad($source_string, $this->array_length, $pad_string, STR_PAD_RIGHT);
        $result = bindec($protected_binary);
        return $result;
    }

    /**
     * Convert the internal stored field array into a binary string representation
     *
     * @param $array_values array The array of values to be converted to a binary string
     * @return string binary representation of the passed boolean array
     * @throws BinaryTranscoderException when the passed array of values is of different length than the instantiated key array
     */
    private function convertArrayToString(array $array_values): string
    {
        array_walk(
            $array_values,
            function (&$item) {
                $item = ($item) ? '1' : '0';
            }
        );
        $result =  \implode($array_values);
        if ($this->array_length != strlen($result)) {
            throw new BinaryTranscoderException('The source array has a different length than the output string!');
        }
        return strrev($result);
    }

    /**
     * Convert a binary string to an array using the internal field array as the new key values
     *
     * @param $source_string string The binary source string to convert
     * @return array
     * @throws BinaryTranscoderException when the string length does not match the array length
     */
    private function convertStringToArray(string $source_string): array
    {
        $key_array = $this->key_array;
        $value_array = str_split($source_string);
        array_walk(
            $value_array,
            function (&$item) {
                $item = (bool)$item;
            }
        );
        if (count($key_array) < count($value_array)) {
            throw new BinaryTranscoderException(
                'The key array does not have enough elements for the decoded integer'
            );
        }
        //WD:RWH - 2018-10-09: this is where the trimming of the value array will be needed to match the key array
        if (null === $this->pad_boolean) {
            $value_array = \array_merge(
                $value_array,
                array_fill(count($value_array), count($key_array)-count($value_array), null)
            );
        }
        if (count($key_array) !== count($value_array)) {
            throw new BinaryTranscoderException(
                'The passed string does not have enough elements for the key array'
            );
        }
        $result = array_combine($key_array, $value_array);
        return $result;
    }

    /**
     * Determine the maximum amount of binary switches that can be stored with the passed
     * integer value.  The default value is PHP_INT_MAX, but if you are using a database to
     * store this information you need to use the maximum value that can be stored in the
     * column that the data is being stored in.
     *
     * @param int $int_max
     * @return int
     */
    public static function determineMaxArrayLength($int_max = PHP_INT_MAX): int
    {
        return strlen(decbin($int_max)) - 1;
    }
}
